### ISSUE TITLE
tracking issue for JAX support in SciPy

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#22246]
_This is part of all the work on array types support (xref gh-18286). JAX support, including `jax.jit` and GPU support, is a multi-step process that is slightly more involved than CuPy/PyTorch support because JAX's programming model diverges more from NumPy, and after the discussion in gh-22070 a separate tracking issue seems warranted._

The full set of changes that are currently on our radar as needed for solid JAX support, from beginning to end, are (not necessarily exactly in order):

- [x] The initial work in gh-20085 that introduced eager JAX support
- [x] The changes to test skip syntax that that helped trigger
- [x] A change to deal with in-place operations with integer indexing (done in [gh-22070](https://github.com/scipy/scipy/pull/22070))
- [ ] A similar change for dealing with in-place operations with boolean masks
- [ ] A way to avoid `jax.jit` choking on input argument validation (xref [array-api-compat#225](https://github.com/data-apis/array-api-compat/issues/225))
- [ ] Making code that calls into compiled code in SciPy work with `jax.jit` (doable as long as shape and dtype of the returned array from the compiled function is predictable, probably with `jax.pure_callback`)
- [ ] Adding a CI job that verifies that `jax.jit` works on SciPy functionality supporting it.

What I think we should do is make these changes as quickly as we can on *a representative subset of SciPy only*, to avoid having a large amount of churn for one of the steps 3-6, across the whole code base and then figuring out that the end goal is somehow blocked by an unforeseen issue. After discussion with @crusaderky (who is doing the heavy lifting at the moment), I think `cluster` plus the part of `special` that already delegates to JAX for matching functions is suitable for this.

We need an assessment at the end of getting `jax.jit` to work on `cluster` and `special`, to decide that we're happy with what JAX support looks like. We expect that the benefits of supporting JAX will be worth the costs - before making a large amount of changes to all other relevant SciPy submodules, let's validate that. The benefits should include:

1. significant enough performance gains of `jax.jit` vs. `numpy` on CPU and GPU,
2. enough of SciPy's functionality/submodules able to have support for `jax.jit` (and `vmap`, ...),

The potential costs are:

1. impact on support for other libraries because of having to rewrite some code constructs in a different style - ensure that there is no significant impact.

    - Expected outcome: No degradation in NumPy (and CuPy and eager PyTorch) performance, aside from perhaps a small O(1) overhead for extra checks - but certainly no extra memory copies or worse asymptotic behavior.
 
2. impact on overall maintainability of the submodules that will gain JAX support

    - Expected outcome: code readability/maintainability takes a small hit, but should not be too bad. The change for in-place operations, replacing `x[idx] += y` by `x = [xpx.at](x)[idx].add(y)` is probably the main change for JAX support. It's annoying unfortunately, but worth trying - and it's hopefully fixable in the future, so we can reintroduce `+=` (there's a plan at least, but it's a lot of work and probably needs >1.5 years, so for now `at` it is)

* * *

It will only be possible to do benchmarking on a representative set of functions towards the end of introducing JAX support, but here is already a taster from a single function (from [this comment](https://github.com/scipy/scipy/pull/20085#issuecomment-2119307509)):

```python
from functools import partial
from typing import Any

import jax
import jax.numpy as jnp
import numpy as np
from scipy import cluster, datasets
import matplotlib.pyplot as plt

show_plot = True

Array = Any  # type this properly later


if __name__ == '__main__':
    # Load an RGB image of a raccoon, of shape (768, 1024, 3)
    face_orig = datasets.face()
    
    # Reshape to 2-D, with each pixel an "observation"
    face = jnp.reshape(face_orig, (-1, 3))  # shape (786432, 3)
    # Dtype is uint8 - use `float32` instead, support in `vq`
    face = face / jnp.float32(255)

    n_clusters = 8
    codebook, _ = cluster.vq.kmeans(face, n_clusters)

    # Apply VQ to map each pixel to the codebook
    quantized, dist = cluster.vq.vq(face, codebook)

    # Reshape quantized data into the original image format
    quantized_image = jnp.reshape(codebook[quantized], face_orig.shape)


    # Plot with Matplotlib
    if show_plot:
        fig, (ax1, ax2) = plt.subplots(1, 2)
        fig.suptitle(f'Original (left) and quantized (right, {n_clusters} colors) raccoons')
        ax1.imshow(face_orig)
        ax2.imshow(quantized_image)

        plt.show()

```

```python
whiten_jit(face).block_until_ready()  # do the JIT compilation
face_gpu = face  # JAX defaults to GPU if that is available
face_cpu = jax.device_put(face, jax.devices('cpu')[0])
face_np = np.asarray(face)

%timeit cluster.vq.whiten(face_np)   #  22 ms
%timeit cluster.vq.whiten(face_cpu)  #   6 ms
%timeit whiten_jit(face_cpu)         #   3 ms
%timeit cluster.vq.whiten(face_gpu)  # 700 us
%timeit whiten_jit(face_gpu)         # 275 us
```

**Comment by mdhaber** [comment]
But there are existing legitimate uses of `xp.where`, so we don't want to mix everything together. 
It sounds like we are tempted to translate `x[idx] = value` to `xpx.at(x, idx).set(value)` in some cases and `x = xp.where(idx, value, x)` (or similar) in others for the sake of immutable arrays. By why not make `xpx.at.set` make all the decisions about which of the three things to do with the underlying array type?

**Comment by rgommers** [comment]
`where` is item 4 on the checklist in the issue description, it isn't actionable yet. It may end up being folded into `at`, or not. It's not a mechanical translation either way, because `where` is out-of-place while `at` is in-place for numpy, so I'd prefer to not dive into this topic just yet.

**Comment by lucascolley** [comment]
right, but we are somewhat diving into it at https://github.com/scipy/scipy/pull/22256#discussion_r1904297542

**Comment by crusaderky** [comment]
> right, but we are somewhat diving into it at [#22256 (comment)](https://github.com/scipy/scipy/pull/22256#discussion_r1904297542)

That specific instance you mention got no benefit from being in-place though

**Comment by mdhaber** [comment]
> That specific instance you mention got no benefit from being in-place though

Great; if that's the case, the change to `where` is fine with me. Just show the evidence for the assertion and I'll be happy to sign off.

**Comment by crusaderky** [comment]
@mdhaber here's a benchmark. It shows that on numpy, `a = np.where(mask, b, a)` is *always* faster than `a[mask] = b[mask]`.

```python
N = 10_000_000  # 80 MB
P = 0.05
import numpy as np
a = np.random.random(N)
b = a + 1
mask = a < P
%timeit a[mask] = b[mask]
30.3 ms ± 185 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit np.where(mask, b, a)
27.1 ms ± 417 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

```python
N = 10_000_000  # 80 MB
P = 0.5
...
%timeit a[mask] = b[mask]
93.5 ms ± 146 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit np.where(mask, b, a)
42 ms ± 82.6 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

```python
N = 100_000  # 800 kB
P = 0.05
...
%timeit a[mask] = b[mask]
192 μs ± 1.99 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit np.where(mask, b, a)
96.1 μs ± 597 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

```python
N = 100_000  # 800 kB
P = 0.5
...
%timeit a[mask] = b[mask]
874 μs ± 19.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit np.where(mask, b, a)
338 μs ± 6.99 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by crusaderky** [comment]
Same benchmark, but for `a[mask] = scalar` vs. `a = np.where(mask, scalar, a)` shows that the update is slighlty faster. This is hardly surprising considering that this time the update doesn't involve a temporary deep copy of rhs.
```python
import numpy as np
N = 10_000_000  # 80 MB
P = 0.05
a = np.random.random(N)
mask = a < P
%timeit a[mask] = 1.0
12.3 ms ± 157 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit np.where(mask, 1.0, a)
18.1 ms ± 235 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by mdhaber** [comment]
Great. I tested locally and mostly reproduced, but showed that when the total number of replacements is small, mutation can be significantly faster on my machine.

```python3
N = 10_000_000  # 80 MB
P = 0.01
import numpy as np
a = np.random.random(N)
b = a + 1
mask = a < P
%timeit a[mask] = b[mask]
8.8 ms ± 246 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit np.where(mask, b, a)
25.1 ms ± 389 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

But yeah, I think the takeaway is that it's not right to assume that mutation is faster. As a rule of thumb, I'd suggest that mutation only be used when the replacement is the exceptional case, otherwise use `where`.

In `logsumexp`, I don't think that replacement is an exceptional case, and I don't think it matters much in general, so I'm OK with replacing that one with `where` with a comment.

**Comment by rgommers** [comment]
That's definitely a little surprising to me, I had expected inplace updates to be more performant. Thanks for putting the benchmark together @crusaderky. That does make things easier, the rule of thumb works for me.

**Comment by ilayn** [comment]
Here is the ballpark of performance when advanced index removed 

```python
N = 10_000_000  # 80 MB
P = 0.05
import numpy as np
a = np.random.random(N)
b = a + 1
mask = a < P
%timeit a[mask] = b[mask]
%timeit np.where(mask, b, a)
%timeit a[:N//2] = b[::2]
23.1 ms ± 631 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
27.7 ms ± 1.27 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
5.7 ms ± 28.3 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by rgommers** [comment]
> Here is the ballpark of performance when advanced index removed

Yes, that is clear. This benchmark and discussion is specifically about using a boolean mask, not slices or integer arrays. For the latter, that's already taken care of (the new `at` functionality will use in-place operations for NumPy et al., and out-of-place for JAX).

**Comment by ilayn** [comment]
True but this functionality requires a true in-place operation which is needed in many parts of the codebase and maybe NumPy should reconsider bool mask handling in a different fashion. 

Because masks are not utilized in a performant manner as the benchmarks above show. So it shouldn't be limited to current availability. This gets less performant with nd-masks.

**Comment by crusaderky** [comment]
This is a wrap-up post on the recent effort to run scipy on JAX:

- Added shims ([xpx.at](https://data-apis.org/array-api-extra/generated/array_api_extra.at.html), [array_api_compat.is_writeable_array](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_writeable_array)) to work around the lack of mutability in JAX
- Added shims ([xpx.lazy_apply](https://data-apis.org/array-api-extra/generated/array_api_extra.lazy_apply.html)) to run Cython kernels in JAX CPU, with the JIT enabled. Before, they would only work in eager mode (as the previous pattern was `xp.asarray(f(np.asarray(x))`). `xpx.lazy_apply` is built on top of `jnp.pure_callback`, which while it carries a substantial performance cost it allows delaying a Cython-less reimplementation of non-critical components.
- Implemented a pattern for input validation on `jax.jit`, where you can't call `bool()` and can't raise exceptions ([array_api_compat.is_lazy_array](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_lazy_array), [scipy._lib._array_api.eager_warns](https://github.com/scipy/scipy/blob/f8ca671a45e2ae9960a7faf9ad89ba45db02e3d5/scipy/_lib/_array_api.py#L265))
- Added infrastructure ([xpx.lazy_xp_function](https://data-apis.org/array-api-extra/generated/array_api_extra.testing.lazy_xp_function.html), [scipy._lib._array_api.xp_capabilities](https://github.com/scipy/scipy/blob/f8ca671a45e2ae9960a7faf9ad89ba45db02e3d5/scipy/_lib/_array_api.py#L780), [jax_autojit](https://github.com/data-apis/array-api-extra/blob/27fbd9c6262ed7c547f17e247e9e21eaddf89576/src/array_api_extra/_lib/_utils/_helpers.py#L530)) to transparently test scipy functions inside `jax.jit`
- Added infrastructure ([scipy._lib._array_api.xp_capabilities](https://github.com/scipy/scipy/blob/f8ca671a45e2ae9960a7faf9ad89ba45db02e3d5/scipy/_lib/_array_api.py#L780)) to automatically generate documentation for the functionality of each function (JAX yes/no, GPU+CPU/only CPU, jittable/only eager), which is by construction coupled with the unit tests and as such can't fall out of sync. [Example here](https://scipy.github.io/devdocs/reference/generated/scipy.cluster.hierarchy.linkage.html) (scroll down to the notes).
- Fully migrated `scipy.cluster`, a Cython-heavy module, to use all of the above.
- Migrated all xp-compatible functions in `scipy.stats`, `scipy.special`, and `scipy.constants` to use all of the above. If a function is compatible with alternative backends, then its level of compatibility with JAX is now both fully documented and fully tested 

## Follow-ups
- There is [ongoing effort](https://github.com/scipy/scipy/issues/22500) to migrate scipy.spatial to JAX. The module, which was originally written in Cython, is being reimplemented to have Array API-agnostic paths and reuses all of the tooling and patterns established above. Early benchmarks show between 10x and 100x performance increase between numpy and JAX GPU.
- There is [ongoing effort](https://github.com/scientific-python/spatch/pull/3) to establish a more robust and unified dispatch system, which will automatically dispatch all `scipy` functionality that has been implemented in `jax.scipy`. This will replace the current ad-hoc system.
- There is [discussion](https://github.com/data-apis/array-api-extra/issues/303) about supporting converging algorithms
- There is discussion ([here](https://github.com/data-apis/array-api/issues/845) and [here](https://github.com/data-apis/array-api-compat/issues/331)) to better support read-only backends in the Array API Standard

## Known issues
- There are false positives in the documentation regarding JAX GPU support, where the tests pass but what is actually happening is that the data is being shunted back to host memory and run through a Cython kernel.  This is blocked by https://github.com/jax-ml/jax/issues/25998. CuPy and PyTorch have no such issue due to differences in their transfer guard described in the linked issue.
- `jax.jit` is currently incapable of honouring input->output device propagation, in other words when the input arrays of a jitted scipy function are on a device that is not the current default device. This is because of https://github.com/jax-ml/jax/issues/26000. CuPy also fails to honour this, while PyTorch works.
- `jax.pure_callback` suffers from poor performance - see below.

## Benchmarks
All benchmarks were run on an AMD 32-core Linux desktop mounting a GeForce 3080 RTX.
When running on JAX, the function is wrapped by `jax.jit` and the JIT is already warm.

Here I'm going to show two very different use cases:
 
### `scipy.cluster.vq.whiten`
This function is implemented on top of the Array API. scipy 1.16 introduced two changes specifically for JAX:
- replaced in-place updates with [xpx.at](https://data-apis.org/array-api-extra/generated/array_api_extra.at.html), which on JAX internally dispatches to either `jax.Array.at` (for slice and scalar indices) or to `jnp.where` (for boolean mask indices);
- conditionally disabled `check_finite` with [array_api_compat.is_lazy_array](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_lazy_array) for all lazy arrays (JAX and Dask, at the moment of writing), as it would call `__bool__` which in turn would disrupt graph tracing.

The benchmark below runs with `check_finite=False`, in order to compare like for like. Keep in mind however that the default value is `True` for eager backends and `False` for lazy ones.

![Image](https://github.com/user-attachments/assets/e51989be-74fe-4354-b31b-787233f9a1b8)

Here we can observe:
- for very small inputs, JAX on CPU (jitted, warm jit) is substantially faster than NumPy.
- for medium to large inputs, JAX on CPU is on par with NumPy. This was a bit surprising, as JAX was running on 32 cores while NumPy on just one.
- for medium to large inputs, all GPU backends vastly outperform the CPU backends.

### `scipy.cluster.hierarchy.linkage(method="single")`
This function is implemented in Cython. The way JAX support was introduced here was to wrap the Cython code with [xpx.lazy_apply](https://data-apis.org/array-api-extra/generated/array_api_extra.lazy_apply.html), which internally calls `jax.pure_callback` when it detects a JAX array.

Additionally, there is special logic for lazy backends which postpones input validation, which can raise, to run inside the lazy segment of the algorithm, immediately before invoking the Cython code. This way we prevent triggering potential segfaults or infinite loops in the Cython code caised by bad data.

![Image](https://github.com/user-attachments/assets/901a93aa-00da-49ad-9446-23fc37955632)

This runs exclusively on CPU backends, as "running on GPU" would simply mean moving back and forth the data between host memory and GPU memory.

Here we can observe how the overhead of all backends other than Dask is O(n) instead of O(1), with JAX running at about 52% the speed of NumPy, and PyTorch at about 72%. This is very surprising and warrants further investigation, as even if there was some untowards deep copy of the data (there _shouldn't_ be any) it would not justify the difference: on 312 kilobytes of data, NumPy runs in 3.8s, PyTorch in 5.2s, and JAX in 7.3s.

It's important to note that the design philosophy here is that `xpx.lazy_apply` is a temporary solution, waiting for a proper native reimplementation (or an agnostic one on top of the Array API). This makes sense in real life situations, where the call to the scipy function is part of a much wider workflow that runs end to end in JAX or another Array API backend.

**Comment by crusaderky** [comment]
Update re. the linkage benchmark above:
- PyTorch slowness is caused by `torch.isfinite`, which on CPU is 10x slower than `np.isfinite`;
- JAX slowness is caused by `jax.pure_callback`; to be precise by chaining two calls to it (one to call `pdist` and another to call the `linkage` Cython kernel proper): https://github.com/jax-ml/jax/issues/29322

**Comment by ev-br** [comment]
Looks like the JAX support has been done and has been stable for a while now. Probably nothing left to do here @rgommers ?

One thing which is unique to this issue is benchmarking results, from @crusaderky 's last message: https://github.com/scipy/scipy/issues/22246#issuecomment-2953190180

These benchmarks are very well hidden in this issue already, and closing this issue will bury them for good. In retrospect, it's too bad we did not remember about them for @mtsokol's blog post, https://labs.quansight.org/blog/array-api-meta-blogpost

**Comment by rgommers** [comment]
I think so, seems like support has been stable for quite a while and we're happy with the performance/maintainability tradeoff. If there's anything structural left at this point, that should probably get its own dedicated new issue. 

I think we can find this back if/when needed. Especially the [wrap-up post above](https://github.com/scipy/scipy/issues/22246#issuecomment-2953190180) is useful I think, more than the benchmark - it's a fairly comprehensive summary of what adding JAX support took, which wasn't obvious when this work started.
